logic_tree_tools.py

```diff
@@ -0,0 +1,920 @@
+    ----------
+    logic_tree_pair : CustomLogicTreePair
+        The logic tree pair from which to extract the source branches.
+
+    Returns
+    -------
+    results : dict
+        A dictionary where keys are branch set short names and values are dictionaries of selected branches
+        for each parameter.
+    """
+
+    branch_param_index_to_description = {
+        "CRU": {
+            0: "deformation_model",
+            1: "time_dependence",
+            2: "MFD",
+            4: "moment_rate_scaling",
+        },
+        "HIK": {1: "MFD", 3: "moment_rate_scaling"},
+        "PUY": {3: "moment_rate_scaling"},
+    }
+
+    results = {}
+
+    # Get the branch parameters for the source logic tree
+    source_logic_tree_branch_params = get_source_branch_parameters_and_values(
+        logic_tree_pair.source_logic_tree
+    )
+
+    # Reduce the source logic tree to the highest weighted branches
+    source_logic_tree_highest_weighted_branches = (
+        reduce_logic_tree_to_nth_highest_weighted_branch(
+            logic_tree_pair.source_logic_tree, 1
+        )
+    )
+
+    # Get the short names of the branch sets in the source logic tree
+    source_logic_tree_branch_set_short_names = [
+        x.short_name for x in logic_tree_pair.source_logic_tree.branch_sets
+    ]
+
+    # Initialize the results dictionary with empty dictionaries for each branch set
+    for (
+        source_logic_tree_branch_set_short_name
+    ) in source_logic_tree_branch_set_short_names:
+        results[source_logic_tree_branch_set_short_name] = {}
+
+    # Iterate through each branch set in the source logic tree
+    for branch_set_index, branch_set in enumerate(
+        logic_tree_pair.source_logic_tree.branch_sets
+    ):
+
+        if branch_set.short_name == "SLAB":
+            results["SLAB"] = {
+                "slab_only_branch": [copy.deepcopy(branch_set.branches[0])]
+            }
+
+        num_params = len(source_logic_tree_branch_params[branch_set.short_name])
+
+        # Iterate through each parameter in the branch set
+        for branch_param_idx in range(num_params):
+
+            selected_branches_per_param = []
+
+            # Allow both values at index branch_param_idx and keep everything else as the highest weighted branch
+            possible_values_for_this_param_index = source_logic_tree_branch_params[
+                branch_set.short_name
+            ][branch_param_idx]
+
+            if len(possible_values_for_this_param_index) == 1:
+                continue
+
+            hwb_values = (
+                source_logic_tree_highest_weighted_branches.branch_sets[
+                    branch_set_index
+                ]
+                .branches[0]
+                .values
+            )
+
+            # branch_values_to_find list will be modified for each param value
+            branch_values_to_find = copy.deepcopy(hwb_values)
+
+            for possible_param_value in possible_values_for_this_param_index:
+
```

Comment:
You seem to have a lot single lines broken into their own lines. I'm big fan of breaking code into logical blocks, however, also don't want to overdo it. E.g. there is no need for a new line here imo.

Reply:
Don't worry about fixing this, its just a comment for the future. This is "research" code, so not worth the effort.

Reply:
I removed a few blank lines but didn't try to get them all